internal/aws/containerinsight/utils.go

```diff
@@ -0,0 +1,230 @@
+		prefix = namespace
+	default:
+		log.Printf("E! Unexpected MetricType: %s", mType)
+	}
+	return prefix
+}
+
+func MetricName(mType string, name string) string {
+	return getPrefixByMetricType(mType) + name
+}
+
+func RemovePrefix(mType string, metricName string) string {
+	prefix := getPrefixByMetricType(mType)
+	return strings.Replace(metricName, prefix, "", 1)
+}
+
+func GetUnitForMetric(metric string) string {
+	return metricToUnitMap[metric]
+}
+
+func ConvertToOTLPMetrics(fields map[string]interface{}, tags map[string]string, logger *zap.Logger) pdata.Metrics {
+	md := pdata.NewMetrics()
+	rms := md.ResourceMetrics()
+	rms.Resize(1)
+	rm := rms.At(0)
+
+	var timestamp pdata.Timestamp
+	resource := rm.Resource()
+	for tagKey, tagValue := range tags {
+		if tagKey == Timestamp {
+			timeNs, _ := strconv.ParseUint(tagValue, 10, 64)
+			timestamp = pdata.Timestamp(timeNs)
+			// convert from nanosecond to millisecond (as emf log use millisecond timestamp)
+			tagValue = strconv.FormatUint(timeNs/uint64(time.Millisecond), 10)
+		}
+		resource.Attributes().UpsertString(tagKey, tagValue)
+	}
+
+	ilms := rm.InstrumentationLibraryMetrics()
+
+	metricType := tags[MetricType]
+	for key, value := range fields {
+		metric := RemovePrefix(metricType, key)
+		unit := GetUnitForMetric(metric)
+		switch t := value.(type) {
+		case int:
+			ilms.Append(intGauge(key, unit, int64(t), timestamp))
+		case int32:
+			ilms.Append(intGauge(key, unit, int64(t), timestamp))
+		case int64:
+			ilms.Append(intGauge(key, unit, t, timestamp))
+		case uint:
+			ilms.Append(doubleGauge(key, unit, float64(t), timestamp))
+		case uint32:
+			ilms.Append(doubleGauge(key, unit, float64(t), timestamp))
+		case uint64:
+			ilms.Append(doubleGauge(key, unit, float64(t), timestamp))
+		case float32:
+			ilms.Append(doubleGauge(key, unit, float64(t), timestamp))
+		case float64:
+			ilms.Append(doubleGauge(key, unit, t, timestamp))
+		default:
+			valueType := fmt.Sprintf("%T", value)
+			logger.Warn("Detected unexpected field", zap.String("key", key), zap.Any("value", value), zap.String("value type", valueType))
+		}
+	}
+
+	return md
+}
+
+func intGauge(metricName string, unit string, value int64, ts pdata.Timestamp) pdata.InstrumentationLibraryMetrics {
+	ilm := pdata.NewInstrumentationLibraryMetrics()
+
+	metric := initMetric(ilm, metricName, unit)
+
+	metric.SetDataType(pdata.MetricDataTypeIntGauge)
+	intGauge := metric.IntGauge()
+
+	updateIntDataPoint(intGauge.DataPoints(), value, ts)
+
+	return ilm
+}
+
+func doubleGauge(metricName string, unit string, value float64, ts pdata.Timestamp) pdata.InstrumentationLibraryMetrics {
+	ilm := pdata.NewInstrumentationLibraryMetrics()
+
+	metric := initMetric(ilm, metricName, unit)
+
+	metric.SetDataType(pdata.MetricDataTypeDoubleGauge)
+	doubleGauge := metric.DoubleGauge()
+	dataPoints := doubleGauge.DataPoints()
+	dataPoints.Resize(1)
+	dataPoint := dataPoints.At(0)
+
+	dataPoint.SetValue(value)
+	dataPoint.SetTimestamp(ts)
```

Comment:
It would be good to see both `intGuage` and `doubleGauge` use the same pattern.  I'd suggest either pulling this into an `updateDoubleDataPoint()` helper or inlining the `updateIntDataPoint()` helper since it doesn't seem to be used elsewhere.

Reply:
Yes. I will inline the function.